### DISCUSSION TITLE
USB Device not recognized after debugging stopped

### DISCUSSION THREAD

**Discussion by Jeff-Birt** [earlephilhower/arduino-pico#2645] [Q&A]
When debugging a simple project using Pi Debug probe, I get a 'USB Device not recognized' when I click the icon to stop debugging. I have to power cycle the Pico to resolve this issue. Was trying this yesterday and work and having a fit trying to figure out what was going on as IT has our computers so screwed w.r.t. USB devices I had to use Zadig to reinstall the Pico driver each time. Testing at home this morning reveled the bug with debugging quickly. Debugging with VS Code and Pico SDK does not produce this error.

**Comment by Jeff-Birt** [comment]
The USB device what was getting goofed up shows up as 'Pico' in Device Manager. Everything works fine on my computer at home, so problem is related to how IT has screwed up our computers at work (I work in EE dept of a university and recent IT changes are causing lots of problems.) After reading the docs on the serial ports this morning I realized that the Pico's virtual serial port is not being enumerated and that this is causing numerous issues.

**Comment by Andy2No** [comment]
@Jeff-Birt, are you allowed to uninstall drivers, in device Manager, at work?

This won't be directly relevant, since I'm using Windows 7 but see my reply in the second post here:
https://github.com/earlephilhower/arduino-pico/discussions/2656

I'm having to uninstall one driver to get the other to work, and found manually putting it into boot/uf2 mode helps, to do the upload.

You can avoid the Zadig stage by using a .inf file (no other parts needed) to specify the driver for the COM port (attached - one for the Pi Pico and one for the Pi Pico 2).

[pico and pico 2 serial infs.zip](https://github.com/user-attachments/files/17997141/pico.and.pico.2.serial.infs.zip) 

These have comments saying they're for Windows 7 or 8, but I don't know if that matters.  Maybe there's a version specific to Windows 10 or 11, somewhere.

There's a trick to getting Windows to accept the new driver, instead of insisting the best one is already installed.  It's safer to use this method than Zadig, because you're not as likely to set a new driver for the wrong device, but a bit more tedious.

In Windows 7, it's roughly:
(right click in Device Manager) Update Driver Software
Browse my computer for...
Let me pick from a list of...
(choose) All devices
(choose) Have Disk
(click) Browse ..
(point it to the relevant serial.inf).

**Reply by Jeff-Birt** [reply]
On work on a computer, I have no admin access so I cannot uninstall drivers via Device Manager.

**Reply by Andy2No** [reply]
I see, @Jeff-Birt.  That's a shame.  You may still be able to install from a .inf file... possibly.  Right clicking on one and choosing "Install" sometimes works, though you may still have to jump through a few hoops to get past Windows telling you the best driver is already installed, and those might or might not be an  acceptable .inf file for Windows 10 or 11.  If not, the newer one is out there somewhere.

Uninstalling a driver is trickier though, with out being able to use Device Manager.  There may still be other ways to do it if Zadig is able to get past whatever they've done.

If you simply don't get to see the device manager, you could try running it by name, as devmgmt.msc, like this:

https://www.dell.com/support/kbdoc/en-uk/000123570/how-to-access-device-manager

- that covers doing it from the Run box or the Command prompt.

**Reply by Jeff-Birt** [reply]
I read the other thread you linked to, and it finally occurred to me that I was supposed to use Zadig to install two separate drivers. I don't recall seeing that in the docs anywhere. I have applied to the IT overlords to temporary admin access to see if I can get it sorted on my work computer.

**Reply by Andy2No** [reply]
@Jeff-Birt, I hope that works out.  I only "needed" Zadig to change the COM port driver to CDC (using the .inf file is an alternative to that).  The removable device one, for .UF2 boot mode, installs automatically, for me, but only after uninstalling the COM port one.  I also ticked the box to delete that driver, at times, not just uninstall it, but that may not be a necessary step, and the rules may be a little bit different for Windows 10/11.

Since you're starting from a messed up state, if you get the option to delete a driver choose that, at least the first time.  If your version of Windows has the option to show unconnected devices in Device Manager, choose that and uninstall any that look likely.  Windows is quite good at installing devices again, even after deleting the drivers, so there's rarely any harm in doing that.

I found USBDeview, by nirsoft, to be quite useful for listing junk disconnected drivers, and it can be used to uninstall them too: https://www.nirsoft.net/utils/usb_devices_view.html (free for personal use, IIRC).  It's not essential, but it is quite handy.

Mostly, it could take some persistence to find a way of doing it that works for your OS and circumstances, but the elements of it will probably be the same.  I went over it all quite a few times before I came up with a workable solution without too many steps.

FWIW, I'm finding that setting the board type to Pimoroni Tiny2350 works better, for both the Pi Pico 2 and the Tiny2350.  The difference is that when I uninstall the COM port driver, I immediately get the removable device .uf2 one on the next step (e.g. after unplugging /  plugging back in with the Boot button held down), instead of having to also uninstall a mystery failed device first.  I haven't mentioned that because it just adds more complication to the description, and may only apply to Windows 7, but it might be worth bearing in mind.

The Pimoroni Tiny2350 is mostly a Pico2 with less GPIO pins wired to header pins, plus it has a GP29/ADC3 pin that the Pico and Pico2 lack.  There's also a Pimoroni Tiny2040, with the same relationship to the Pi Pico RP2040.  In most cases, they'll run the same code and it shouldn't matter much which variant of board based on the same chip you set in Tools->Board, except maybe for default pin numbers for things like serial IO and SPI - which you can always specify in the code.  It does seem Pimoroni have done their drivers differently, somehow, so that may be worth a try.

Basically, go over it bearing these things in mind, making notes of what you've tried.  If you get to a stage where it works well enough and is repeatable, you can maybe then refine it some more.

**Comment by Jeff-Birt** [comment]
After getting temp admin privileges back I was able to change the USB driver for 'Pico' to the USB Serial (CDC) and now things seem to work OK. If I power on with the BOOTSEL button down I get 'RP2 BOOT' and RPI RP2 USB Device showing up in Device Manger. In this state I can program from the Arduino IDE w/Pico core or VS Code with the Pico SDK. Once programmed with the Arduino IDE w/Pico core the virtual serial port (COM7 in my case) shows up. This allows for reprogramming without a power cycle. Thanks for all your help.

**Comment by Andy2No** [comment]
You're welcome, Jeff.  I'm glad you've got it to work.

EDIT: For anyone else experiencing similar problems with an RP2350 based board (RPI Pico 2, Pimoroni Plus 2 etc) not going into .uf2 drive mode or not showing up as a serial port, the other related discussion is #2656
